### PR TITLE
[torch.fx] Add support for custom op

### PR DESCRIPTION
**Pull request by houseroad**
Summary: when using torch.fx to symbolic trace nn module, we create Proxy node instead of real Tensor, and Proxy has implemented '__torch_function__'. However, the C++ custom op is just cpp_function. In current cpp_function definition, there is no support for '__torch_function__' override, so we will see complaints about the C++ function expect Tensor input, but got Proxy type. So when calling the cpp_function of the custom op, we first check whether the input args support '__torch_function__' override, if so, we will try to call the override logic instead of the real C++ custom op.

Some perf benchmark:
- sigmoid (input is torch.ones(2))
without change: Op sigmoid had a minimum time of 2.506256103515625 us and a standard deviation of 0.3828478220384568 us.
with change: Op sigmoid had a minimum time of 2.7022361755371094 us and a standard deviation of 0.3112237900495529 us.
- relu (input is torch.ones(2))
without change: Op relu had a minimum time of 2.2428035736083984 us and a standard deviation of 0.4182197153568268 us.
with change: Op relu had a minimum time of 2.4945735931396484 us and a standard deviation of 0.3914815024472773 us.
- cat (input is torch.ones(2), torch.zeros(2))
without change: Op cat had a minimum time of 3.509521484375 us and a standard deviation of 0.5249808309599757 us.
with change: Op cat had a minimum time of 4.532814025878906 us and a standard deviation of 0.6780577823519707 us.

- cat (input is torch.ones(3, 224, 224), torch.zeros(3, 224, 224))
without change: Op cat had a minimum time of 107.0699691772461 us and a standard deviation of 6.263310555368662 us.
with change: Op cat had a minimum time of 106.03904724121094 us and a standard deviation of 5.618293769657612 us.

Test Plan: buck test //caffe2/test:fx -- test_torch_custom_ops

Differential Revision: D23203204

### CODE REVIEW

File: torch/csrc/utils/python_arg_parser.h [hunk 1]
```diff
@@ -769,13 +769,24 @@ static auto check_has_torch_function(PyObject* obj) -> bool
...
+ *
+ * 'overloaded_args' is the args which have overloaded __torch_function__.
+ *
+ * 'func_name' is the named of the original torch method.
  */
 // Used for Tensor methods with arguments.
 auto handle_torch_function(PythonArgs &r, PyObject* self, PyObject* args, PyObject* kwargs, PyObject* torch_api, const char* module_name) -> PyObject*;
 
-// Used fpr functions.
+// Used for fpr functions which needs to parse python args.
```
Comment (hameerabbasi) [inline torch/csrc/utils/python_arg_parser.h:784]:
```suggestion
// Used for functions which needs to parse python args.
```

File: torch/csrc/utils/python_arg_parser.h [hunk 2]
```diff
@@ -769,13 +769,24 @@ static auto check_has_torch_function(PyObject* obj) -> bool
...
 auto handle_torch_function(PythonArgs &r, PyObject* args, PyObject* kwargs, PyObject* torch_api, const char* module_name) -> PyObject*;
 
+// Used for fpr functions created in C++, e.g., C++ custom op, whose args are already parsed.
```
Comment (hameerabbasi) [inline torch/csrc/utils/python_arg_parser.h:787]:
```suggestion
// Used for functions created in C++, e.g., C++ custom op, whose args are already parsed.
```

File: torch/csrc/jit/python/init.cpp [hunk 3]
```diff
@@ -885,13 +886,107 @@ void initJITBindings(PyObject* module) {
...
+                    } else if (kwargs.contains(
+                                   expected_args[i].name().c_str())) {
+                      const auto& kwarg =
+                          kwargs[expected_args[i].name().c_str()];
+                      if (THPVariable_CheckExact(kwarg.ptr())) {
+                        continue;
+                      }
+                      if (check_has_torch_function(kwarg.ptr())) {
+                        append_overloaded_arg(overloaded_args, kwarg.ptr());
+                      }
```
Comment (hameerabbasi) [inline torch/csrc/jit/python/init.cpp:945]: This logic is flawed and will cause the same issue as in #34294. We need a way to recurse into `list`s and `tuple`s.

**Review by hameerabbasi — COMMENTED** [review]
I'm not sure if we plan to support `TensorList`s in this structure, otherwise this seems okay. I've only reviewed the parts pertaining to `__torch_function__`.

File: torch/csrc/jit/python/init.cpp [hunk 4]
```diff
@@ -885,13 +886,121 @@ void initJITBindings(PyObject* module) {
...
+                            PyList_Check(arg)) {
+                          size_t size = PyList_GET_SIZE(arg);
+                          for (size_t idx = 0; idx < size; idx++) {
+                            PyObject* element = PyList_GET_ITEM(arg, idx);
+                            if (check_has_torch_function(element)) {
+                              append_overloaded_arg(overloaded_args, element);
+                            }
+                          }
+                        }
+                        // No tuple type is supported in args and kwargs.
```
Comment (hameerabbasi) [inline torch/csrc/jit/python/init.cpp:947]: IIUC, `TensorList` supports both a tuple and a list of `Tensor`s.

Comment (houseroad) [reply torch/csrc/jit/python/init.cpp]: sure, added the support for tuple and the corresponding test case.

**Comment by houseroad** [comment]
@hameerabbasi could you take another look? @ezyang, could you help on review?

**Review by hameerabbasi — APPROVED** [review]
I've reviewed only the changes pertinent to `__torch_function__`. Thanks a load for all the detailed comments. It looks good to me.

File: torch/csrc/jit/python/init.cpp [hunk 5]
```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
...
+                    if (match) {
+                      found_op = op;
+                      break;
+                    }
+                  }
+                }
+                std::vector<py::handle> overloaded_args;
+                if (found_op) {
+                  auto check_and_append_overloaded_arg =
+                      [](std::vector<py::handle>& overloaded_args,
```
Comment (dzhulgakov) [inline torch/csrc/jit/python/init.cpp:928]: can you add a big fat comment that it needs to be unified in general with logic in python_arg_parser?

File: torch/csrc/jit/python/init.cpp [hunk 6]
```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
...
+          auto func_holder = std::make_unique<py::cpp_function>();
+          py::cpp_function& func_ref = *func_holder;
+          auto func = py::cpp_function(
+              [operations, op_name, self_ptr = std::move(func_holder)](
+                  py::args args, py::kwargs kwargs) {
+                std::shared_ptr<Operator> found_op = nullptr;
+                if (operations.size() == 1) {
+                  found_op = operations.at(0);
+                } else {
+                  for (const auto& op : operations) {
```
Comment (dzhulgakov) [inline torch/csrc/jit/python/init.cpp:908]: this matching is still not ideal - what if there are two overrides:
```
foo(x : Tensor)
foo(x : List[Tensor])
```

Since you don't verify the matching here, it's going to pick the first overload and miss iterating over the list, right?

I don't really get why we need to do schema matching here at all, for a baseline implementation we could just iterate blindly over the entire list of arguments (as you do anyway) and if it's Python's tuple or list - then recurse

File: torch/csrc/jit/python/init.cpp [hunk 7]
```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
...
+                  auto check_and_append_overloaded_arg =
+                      [](std::vector<py::handle>& overloaded_args,
+                         const Argument& expected_arg,
+                         PyObject* arg) {
+                        if (THPVariable_CheckExact(arg)) {
+                          return;
+                        }
+                        if (check_has_torch_function(arg)) {
+                          append_overloaded_arg(overloaded_args, arg);
+                        } else if (expected_arg.type()->cast<ListType>()) {
```
Comment (dzhulgakov) [inline torch/csrc/jit/python/init.cpp:936]: it seems to be the only place where you actually use Schema. Can't you just do the check on `is_tuple || is_list` here and call it a day?

File: torch/csrc/jit/python/init.cpp [hunk 8]
```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
...
+                    std::vector<py::object> overloaded_types;
+                    overloaded_types.reserve(overloaded_args.size());
+                    for (auto& oarg : overloaded_args) {
+                      overloaded_types.push_back(
+                          py::reinterpret_borrow<py::object>(
+                              (PyObject*)Py_TYPE(oarg.ptr())));
+                    }
+                    py::tuple py_types = py::cast(overloaded_types);
+                    py::object ret;
+                    size_t delim_pos = op_name.find("::");
```
Comment (dzhulgakov) [inline torch/csrc/jit/python/init.cpp:978]: You can use `Symbol` for parsing it (better yet, just pass `Symbol` which is just 4 bytes as a capture to this function instead of `op_name`)

There's `Symbol::toUnqualString()` for the second part and `Symbol::ns()::toUnqualString()` for the first

File: torch/csrc/utils/python_arg_parser.h [hunk 9]
```diff
@@ -785,4 +796,46 @@ auto handle_torch_function_getter(THPVariable* self, const std::string& property
...
+ *
+ *  'obj' is an object to check for a __torch_function__ implementation
+ *
+ * If changing this file in a way that can affect the __torch_function__
+ * overhead, please report the benchmarks in 'benchmarks/overrides_benchmark'.
+ * See the instructions in the 'README.md' in that directory.
+ *
+ */
+
+void append_overloaded_arg(std::vector<py::handle> &overloaded_args, PyObject* obj);
```
Comment (dzhulgakov) [inline torch/csrc/utils/python_arg_parser.h:839]: it's kind of hard to get what this function does from the name now that it's public. Maybe `append_torch_function_overloaded_arg`?

File: torch/csrc/jit/python/init.cpp [hunk 10]
```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
...
+                      }
+                    }
+                    if (match) {
+                      found_op = op;
+                      break;
+                    }
+                  }
+                }
+                std::vector<py::handle> overloaded_args;
+                if (found_op) {
```
Comment (dzhulgakov) [inline torch/csrc/jit/python/init.cpp:926]: this function is really hard to parse (3 levels of nested lambdas!) maybe extract at least part related to handling torch function into a subfunction?

File: torch/csrc/utils/python_arg_parser.h [hunk 11]
```diff
@@ -769,13 +769,24 @@ static auto check_has_torch_function(PyObject* obj) -> bool
...
  */
 // Used for Tensor methods with arguments.
 auto handle_torch_function(PythonArgs &r, PyObject* self, PyObject* args, PyObject* kwargs, PyObject* torch_api, const char* module_name) -> PyObject*;
 
-// Used fpr functions.
+// Used for functions which needs to parse python args.
 auto handle_torch_function(PythonArgs &r, PyObject* args, PyObject* kwargs, PyObject* torch_api, const char* module_name) -> PyObject*;
 
+// Used for functions created in C++, e.g., C++ custom op, whose args are already parsed.
+auto handle_torch_function(const std::vector<py::handle> &overloaded_args, PyObject* args, PyObject* kwargs, const char* func_name, PyObject* torch_api_function, const char* module_name) -> PyObject*;
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.h:788]: nit: `ArrayRef<py::handle>` is probably better as it doesn't mandate that you specifically use vector here

File: torch/csrc/utils/python_arg_parser.h [hunk 12]
```diff
@@ -769,13 +769,24 @@ static auto check_has_torch_function(PyObject* obj) -> bool
  *
  * 'torch_api' is a reference to a python torch API namespace.
  *
+ * 'torch_api_function' is the reference to the original torch method, usually,
+ * we can use torch_api and func_name to get torch_api_function. In some cases,
+ * e.g., torch custom op, we create the funciton in C++, if we still use
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.h:774]: function

File: torch/csrc/jit/python/init.cpp [hunk 13]
```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
             docstring << "  " << op->schema() << "\n";
           }
 
-          return py::cpp_function(
-              [operations](py::args args, py::kwargs kwargs) {
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:889]: Just a 10 mile view comment here: this is a lot more code than I would have expected here!

**Comment by ezyang** [comment]
> Summary: First override torch.ops, then export the custom op as call_function. After symbolic execution, restore the original _Ops to torch.ops.

This PR description needs updating. The patch is pretty involved now and it would be really good to have an actual explanation of what is going on in English.

File: torch/csrc/jit/python/init.cpp [hunk 14]
```diff
@@ -885,13 +886,125 @@ void initJITBindings(PyObject* module) {
...
+          // To support __torch_function__ override mechanism, `func` needs to
+          // provide a pointer, which points to the original method, i.e.,
+          // itself. py::cpp_function doesn't support specifying the function
+          // signature without real implementation, so we we use a trick here.
+          // First, we create a function holder pointing to an empty
+          // cpp_function. Then we use a reference as the alias of the empty
+          // function. Then we define the real returned function, and update
+          // the value of the function pointed by the holder using the alias.
+          // With this trick, we are able to call self inside the returned
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:897]: not call, surely? Just reference the `cpp_function`

**Comment by ezyang** [comment]
Can you explain to me why we have to reimplement the overload resolution procedure to implement `__torch_function__` here? That's pretty unexpected.

File: torch/csrc/jit/python/init.cpp (hunk 8 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp]: Good point, forgot to use "Symbol"...

File: torch/csrc/jit/python/init.cpp (hunk 10 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp]: moved it out, and fused with append_overloaded_args

**Comment by houseroad** [comment]
Hi @ezyang, I have updated the summary. In short, custom op is exposed as cpp_function in python, and we don't have specific logic for __torch_function__ override.

File: torch/csrc/jit/python/init.cpp (hunk 13 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp:901]: Sure, refactored a bit, and should be easy to understand now.

Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:901]: OK, this is much better, thank you!

File: torch/csrc/utils/python_arg_parser.h [hunk 15]
```diff
@@ -785,4 +796,19 @@ auto handle_torch_function_getter(THPVariable* self, const std::string& property
...
+ *
+ * Check if the input arg supports __torch_function__, if so, append it to
+ * overloaded_args. If arg is a list or tuple, further check whether each
+ * element in list/tuple supports override.
+ *
+ * 'arg': the input argument to be checked
+ * 'overloaded_args': the vector to append the overloaded args.
+ *
+ */
+void extract_torch_function_overloaded_arg(PyObject* arg, std::vector<py::handle>& overloaded_args);
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.h:812]: Instead of taking the vector in by mutable reference, it will be clearer to take it by pointer, so you get the visual clue at the call site that this vector is going to get mutated.

File: torch/csrc/jit/python/init.cpp [hunk 16]
```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
...
+          // function. Then we define the real returned function, and update
+          // the value of the function pointed by the holder using the alias.
+          // With this trick, we are able to reference self inside the returned
+          // function.
+          auto func_holder = std::make_unique<py::cpp_function>();
+          py::cpp_function& func_ref = *func_holder;
+          auto func = py::cpp_function(
+              [operations, symbol, self_ptr = std::move(func_holder)](
+                  py::args args, py::kwargs kwargs) {
+                std::vector<py::handle> overloaded_args;
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:916]: If I recall correctly, when @hameerabbasi was doing benchmarking, having overloaded argument detection separate from normal argument parsing was a noticeable overhead (that's why it's interleaved with regular argument parsing in the normal implementation). This implementation here is probably OK for now, but if you ever turn around and want to start optimizing the performance of `torch.ops`, this code here is going to make things harder to do so. But it is probably OK to not worry about this too much right now, because there are probably a bunch of performance problems with torch.ops today already, and so if we want to tackle perf here we should actually measure.

Regardless, it might be good to do a little microbenchmark and see exactly how much we're losing here.

cc @ljk53

File: torch/csrc/jit/python/init.cpp [hunk 17]
```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
...
+                for (size_t i = 0; i < args.size(); ++i) {
+                  extract_torch_function_overloaded_arg(
+                      args[i].ptr(), overloaded_args);
+                }
+                // NB: for kwargs, we cannot guarantee the order of appending
+                // is the same as the argument order in operator's schema.
+                // This is suboptimal, but should be fine. Later when we have
+                // better schema matching and argument parsering, we could
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:924]: nit: parsing

File: torch/csrc/jit/python/init.cpp [hunk 18]
```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
...
+                }
+                if (overloaded_args.size() > 0) {
+                  std::vector<py::object> overloaded_types;
+                  overloaded_types.reserve(overloaded_args.size());
+                  for (auto& oarg : overloaded_args) {
+                    overloaded_types.push_back(
+                        py::reinterpret_borrow<py::object>(
+                            (PyObject*)Py_TYPE(oarg.ptr())));
+                  }
+                  py::tuple py_types = py::cast(overloaded_types);
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:939]: Efficiency nit: it would be more efficient to directly put the overloaded types into the python tuple directly, rather than going through a vector

File: torch/csrc/jit/python/init.cpp [hunk 19]
```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
...
+                  py::object ret;
+                  std::string module_name("torch.ops");
+                  module_name.append(symbol.ns().toUnqualString());
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:942]: Don't you need a period to separate "torch.ops" and the namespace?

File: torch/csrc/jit/python/init.cpp [hunk 20]
```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
...
+                  std::string method_name = symbol.toUnqualString();
+                  return pybind11::reinterpret_borrow<py::object>(
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:944]: This looks very suspicious

Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:944]: This is wrong, you need to steal the returned PyObject, as it is an owning reference (as seen by the `release` in the implementation of `handle_torch_function`. If you return a `py::object` directly from `handle_torch_function` that would make it more difficult to make this mistake in the future

File: torch/csrc/jit/python/init.cpp [hunk 21]
```diff
@@ -897,13 +898,65 @@ void initJITBindings(PyObject* module) {
...
+          // To support __torch_function__ override mechanism, `func` needs to
+          // provide a pointer, which points to the original method, i.e.,
+          // itself. py::cpp_function doesn't support specifying the function
+          // signature without real implementation, so we we use a trick here.
+          // First, we create a function holder pointing to an empty
+          // cpp_function. Then we use a reference as the alias of the empty
+          // function. Then we define the real returned function, and update
+          // the value of the function pointed by the holder using the alias.
+          // With this trick, we are able to reference self inside the returned
+          // function.
```
Comment (ezyang) [inline torch/csrc/jit/python/init.cpp:910]: I'm concerned about reference cycle here. `func_holder` holds an owning reference to the (refcounted) `cpp_function` and you then transfer ownership of this unique reference to the `cpp_function`. So `cpp_function` has a owning reference to itself, reference cycle.

What you should probably do instead maintain a non-owning to cpp function inside the cpp function. You are guaranteed that it will be live when you use it in the function because, how else could you have called the function; only when you pass it into a function call do you increase refcount.

cc @dzhulgakov

File: torch/csrc/utils/python_arg_parser.h [hunk 22]
```diff
@@ -769,13 +769,24 @@ static auto check_has_torch_function(PyObject* obj) -> bool
...
+ * 'func_name' is the named of the original torch method.
  */
 // Used for Tensor methods with arguments.
 auto handle_torch_function(PythonArgs &r, PyObject* self, PyObject* args, PyObject* kwargs, PyObject* torch_api, const char* module_name) -> PyObject*;
 
-// Used fpr functions.
+// Used for functions which needs to parse python args.
 auto handle_torch_function(PythonArgs &r, PyObject* args, PyObject* kwargs, PyObject* torch_api, const char* module_name) -> PyObject*;
 
+// Used for functions created in C++, e.g., C++ custom op, whose args are already parsed.
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.h:787]: Technically, they're not already parsed; rather, we're doing overload handling prior to parsing, no?

Also, I feel it would be a lot clearer if we stop overloading the function here, and just gave separate names to each variant. There doesn't really seem to be a good reason to be overloading here.

File: torch/csrc/utils/python_arg_parser.cpp [hunk 23]
```diff
@@ -976,4 +987,28 @@ at::Scalar PythonArgs::scalar_slow(int i) {
   return at::Scalar(THPUtils_unpackDouble(args[i]));
 }
 
+void extract_torch_function_overloaded_arg(PyObject* arg, std::vector<py::handle>& overloaded_args) {
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.cpp:990]: This is a duplicate of code in PythonArgParser and it should be cross-referenced here and there accordingly

**Comment by ezyang** [comment]
This is probably about as good as it is going to get without a serious rewrite of how python argument parsing works in general. There are some correctness issues I raised in review, please take a look.

File: torch/csrc/utils/python_arg_parser.h (hunk 15 above)
Comment (houseroad) [reply torch/csrc/utils/python_arg_parser.h:812]: Sure, i just kept it consistent with append_overloaded_arg defined in python_arg_parser.cpp.

File: torch/csrc/jit/python/init.cpp (hunk 18 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp:939]: Do we have a way to append to tuple?

File: torch/csrc/jit/python/init.cpp (hunk 19 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp:942]: Good catch

File: torch/csrc/jit/python/init.cpp (hunk 20 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp:944]: Good catch. It indeed should be steal instead of borrow. I will fix it, and leave the API change to the following refactoring.

File: torch/csrc/utils/python_arg_parser.h (hunk 22 above)
Comment (houseroad) [reply torch/csrc/utils/python_arg_parser.h:787]: Sure, I will still keep handle_torch_function in this diff, and leave the renaming to the next diff. Otherwise, we will have quite a lot of distractive changes. There are quite a lot uses.

File: torch/csrc/utils/python_arg_parser.cpp (hunk 23 above)
Comment (houseroad) [reply torch/csrc/utils/python_arg_parser.cpp:990]: This logic is a subset in FunctionParameter::check, https://fburl.com/diffusion/vc724gz6. I am not sure how to share the whole 'extract_torch_function_overloaded_arg' with PythonArgParser without large refactoring, since we need to match the schema ahead (we don't have a good way to match the schema yet for custom op in python), then we can try to generate a string fmt to create FunctionParameter. One thing I could do here is lift the logic of handling TENSOR_LIST to a function, so it could be shared here and there.

File: torch/csrc/jit/python/init.cpp (hunk 16 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp:905]: okay, then i will just use py::import to fetch the pointer to original method then.

File: torch/csrc/utils/python_arg_parser.cpp (hunk 23 above)
Comment (houseroad) [reply torch/csrc/utils/python_arg_parser.cpp]: okay, I have merge this logic with the logic in FunctionParameter::check.

File: torch/csrc/jit/python/init.cpp (hunk 21 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp]: okay, i will just use py::import to get the original method.

File: torch/csrc/jit/python/init.cpp (hunk 16 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp:905]: @ezyang do you know how we benchmarked the torch.ops before? also cc: @smessmer

File: torch/csrc/utils/python_arg_parser.h (hunk 22 above)
Comment (ezyang) [reply torch/csrc/utils/python_arg_parser.h]: Sure, just make the new function have a different name

File: torch/csrc/jit/python/init.cpp (hunk 18 above)
Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:939]: You cannot append to a tuple but in the raw Python API you preallocate a tuple of the desired size and then directly set elements at each index. I don't know if pybind11 gives a nice way to do this, you'd have to check their docs.

File: torch/csrc/jit/python/init.cpp (hunk 16 above)
Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:905]: I don't think we've done any torch.ops benchmarking; c.f., https://github.com/pytorch/pytorch/pull/42341 . But I would go about benchmarking them the same way I benchmark normal ops. cc @robieta @ngimel @wconstab

**Review by ezyang — APPROVED** [review]
Thanks!

File: torch/csrc/utils/python_arg_parser.h (hunk 22 above)
Comment (houseroad) [reply torch/csrc/utils/python_arg_parser.h]: Sure, as you wish :-)

File: torch/csrc/jit/python/init.cpp (hunk 18 above)
Comment (houseroad) [reply torch/csrc/jit/python/init.cpp:939]: just realized python tuple is immutable.

**Comment by houseroad** [comment]
The failing test is unrelated, and the benchmarking results are added in the summary.

**Comment by houseroad** [comment]
Benchmark script is added in D23469918

File: torch/csrc/jit/python/init.cpp (hunk 21 above)
Comment (dzhulgakov) [reply torch/csrc/jit/python/init.cpp]: good catch, we could you `py::handle` (the unowning reference). But py::import works too since perf doesn't matter when we trigger `__torchfunction__`
